python/ray/train/utils.py

```diff
@@ -112,3 +112,18 @@ def __call__(cls, *args, **kwargs):
         if cls not in cls._instances:
             cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
         return cls._instances[cls]
+
+
+class ResultsList(list):
```

Comment:
It is a little bit complicated. There was an agreement on a temporary solution (by appending aggregated metrics to each result dict first, previous comments [1](https://github.com/ray-project/ray/pull/22099#pullrequestreview-890618542), [2](https://github.com/ray-project/ray/pull/22099#issuecomment-1049138730).) Some considerations are
1. We don't want to directly append aggregated metrics to `results_list`. This will change `results_list` (for instance, increase its length by 1) and causes test failures.
2. (1) or (2) in the comment 1 essentially suggests appending aggregated metrics to all result dicts, since we don't know which worker will be logged. One drawback of this is duplicated aggregate metrics in `results_list`.
3. The `ResultsList` class is created for keeping the existing properties of `results_list` ad avoiding duplicate aggregated metrics. 

Maybe I misunderstood your previous comment. Now reflecting on what you suggested last Friday, I think what you meant is changing each worker's result dict into a nested dict, i.e. `results_list = [dict_worker_0, ..., dict_worker_n]` where `dict_worker_0 = {"worker_0": existing_metrics, "avg": average_metrics, "max": max_metrics,...}`. Is it correct?

Reply:
Ok that's fine, we can leave this for now as an internal abstraction, but I would rather not expose this to users, even those who are implementing custom callbacks or custom preprocessors.